File: 1214-two-sum-bsts/1214-two-sum-bsts.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def dfs(self, root, i):
        if root is None:
            return
        self.dfs(root.left, i)
        self.nums[i].append(root.val)
        self.dfs(root.right, i)
    def twoSumBSTs(self, root1, root2, target):
        """
        :type root1: TreeNode
        :type root2: TreeNode
        :type target: int
        :rtype: bool
        """


        self.nums = [[], []]
        self.dfs(root1, 0)
        self.dfs(root2, 1)
        i, j = 0, len(self.nums[1]) - 1
        while i < len(self.nums[0]) and ~j:
            x = self.nums[0][i] + self.nums[1][j]
            if x == target:
                return True
            if x < target:
                i += 1
            else:
                j -= 1
        return False

File: 1214-two-sum-bsts/1214-two-sum-bsts.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<Integer> nums1 = new ArrayList<>();
    private List<Integer> nums2 = new ArrayList<>();

    private void inorder(TreeNode root, List<Integer> nums) {
        if (root == null) {
            return;
        }
        inorder(root.left, nums);
        nums.add(root.val);
        inorder(root.right, nums);
    }

    public boolean twoSumBSTs(TreeNode root1, TreeNode root2, int target) {
        inorder(root1, nums1);
        inorder(root2, nums2);
        int i = 0;
        int j = nums2.size() - 1;
        while (i < nums1.size() && j >= 0) {
            int sum = nums1.get(i) + nums2.get(j);
            if (sum == target) {
                return true;
            }
            if (sum < target) {
                i++;
            } else {
                j--;
            }
        }
        return false;
    }
}